File: network/src/main/java/com/github/vivchar/network/StargazersManager.java
package com.github.vivchar.network;

import android.support.annotation.NonNull;
import android.util.Log;

import com.github.vivchar.network.models.GithubUser;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Observable;
import io.reactivex.functions.Function;
import io.reactivex.subjects.ReplaySubject;


/**
 * Created by Vivchar Vitaly on 09.10.17.
 */

public class StargazersManager {

	private static final String TAG = StargazersManager.class.getSimpleName();

	@NonNull
	private final GithubClient mClient;
	@NonNull
	private final ReplaySubject<List<GithubUser>> mGithubUsers = ReplaySubject.createWithSize(1);

	private int mCurrentPage = 1;
	private int mLastLoadedPage = 0;
	private boolean mHasMore = false;
	private boolean mReloading = false;
	@NonNull
	private final List<GithubUser> mOriginalUsers = new ArrayList<>();

	StargazersManager(@NonNull final GithubClient client) {
		mClient = client;
	}

	public void sendReloadRequest() {
		Log.d(TAG, "sendReloadRequest");
		/* vivchar: to avoid the API rate limit of the github https://developer.github.com/v3/#rate-limiting */
		if (mGithubUsers.getValue().isEmpty()) {
			mReloading = true;
			mCurrentPage = 1;
			sendPageLoadRequest(mCurrentPage);
		} else {
			mGithubUsers.onNext(new ArrayList<>(mOriginalUsers)); //temporary workaround
		}
	}

	public void sendLoadMoreRequest() {
		if (mHasMore) {
			sendPageLoadRequest(mLastLoadedPage + 1);
		}
	}

	private void sendPageLoadRequest(final int page) {
		Log.d(TAG, "sendPageLoadRequest: " + page);
		mClient.sendStargazersRequest(page);
	}

	public void onApplicationStarted() {
		sendPageLoadRequest(mCurrentPage);
	}

	public void onStargazersReceived(final int page, @NonNull final List<GithubUser> list) {
		Log.d(TAG, "onStargazersReceived: " + page + " list: " + list.size());
		mLastLoadedPage = page;
		mHasMore = !list.isEmpty();

		final ArrayList<GithubUser> newList = new ArrayList<>();
		if (mReloading) {
			mReloading = false;
		} else if (mGithubUsers.getValue() != null) { //load more response
			newList.addAll(0, mGithubUsers.getValue());
		}
		newList.addAll(list);
		mOriginalUsers.addAll(new ArrayList<>(newList));

		mGithubUsers.onNext(newList);
	}

	public void onStargazersFailed(final int page) {
		Log.e(TAG, "onStargazersFailed: " + page);
		mGithubUsers.onNext(new ArrayList<>());
	}

	@NonNull
	public Observable<List<GithubUser>> getAll() {
		return mGithubUsers.hide();
	}

	@NonNull
	public Observable<List<GithubUser>> getTop10() {
		return mGithubUsers.hide().map(githubUsers -> new ArrayList<>(githubUsers.subList(0, Math.min(githubUsers.size(), 10))));
	}
}

File: network/src/main/java/com/github/vivchar/network/StargazersManager.kt
package com.github.vivchar.network

import android.util.Log
import com.github.vivchar.network.models.GithubUser
import io.reactivex.Observable
import io.reactivex.subjects.ReplaySubject
import java.util.*

/**
 * Created by Vivchar Vitaly on 09.10.17.
 */
class StargazersManager internal constructor(private val client: GithubClient) {

	private val githubUsersSubject = ReplaySubject.createWithSize<List<GithubUser>?>(1)
	private var currentPage = 1
	private var lastLoadedPage = 0
	private var hasMore = false
	private var isReloading = false
	private val originalUsers: MutableList<GithubUser> = ArrayList()

	fun sendReloadRequest() {
		Log.d(TAG, "sendReloadRequest")
		/* vivchar: to avoid the API rate limit of the github https://developer.github.com/v3/#rate-limiting */
		if (githubUsersSubject.value!!.isEmpty()) {
			isReloading = true
			currentPage = 1
			sendPageLoadRequest(currentPage)
		} else {
			githubUsersSubject.onNext(ArrayList(originalUsers)) //temporary workaround
		}
	}

	fun sendLoadMoreRequest() {
		if (hasMore) {
			sendPageLoadRequest(lastLoadedPage + 1)
		}
	}

	private fun sendPageLoadRequest(page: Int) {
		Log.d(TAG, "sendPageLoadRequest: $page")
		client.sendStargazersRequest(page)
	}

	fun onApplicationStarted() {
		sendPageLoadRequest(currentPage)
	}

	fun onStargazersReceived(page: Int, list: List<GithubUser>) {
		Log.d(TAG, "onStargazersReceived: " + page + " list: " + list.size)
		lastLoadedPage = page
		hasMore = list.isNotEmpty()
		val newList = ArrayList<GithubUser>()
		if (isReloading) {
			isReloading = false
		} else if (githubUsersSubject.value != null) { //load more response
			newList.addAll(0, githubUsersSubject.value!!)
		}
		newList.addAll(list)
		originalUsers.addAll(ArrayList(newList))
		githubUsersSubject.onNext(newList)
	}

	fun onStargazersFailed(page: Int) {
		Log.e(TAG, "onStargazersFailed: $page")
		githubUsersSubject.onNext(ArrayList())
	}

	val all: Observable<List<GithubUser>> get() = githubUsersSubject.hide()

	val top10: Observable<List<GithubUser>> get() = githubUsersSubject.hide().map { ArrayList(it.subList(0, it.size.coerceAtMost(10))) }

	companion object {
		private val TAG = StargazersManager::class.java.simpleName
	}
}